tests/local/component-model/type-export-restrictions.wast

```diff
@@ -0,0 +1,427 @@
+  (func $f (param "f" $t) (canon lift (core func $i "f")))
+  (export "f" (func $f))
+)
+
+;; a type-ascribed export which is otherwise invalid can become valid
+(component
+  (type $t (record))
+
+  (core module $m (func (export "f")))
+  (core instance $i (instantiate $m))
+  (func $f (param "x" $t) (canon lift (core func $i "f")))
+
+  (export $t' "t" (type $t))
+  (export "f" (func $f) (func (param "x" $t')))
+)
+
+;; imports can't reference exports
+(assert_invalid
+  (component
+    (type $t1 (record))
+    (export $t2 "t1" (type $t1))
+    (import "i" (func (result $t2)))
+  )
+  "func not valid to be used as import")
+
+;; exports can reference imports
+(component
+  (type $t1 (record))
+  (import "t1" (type $t2 (eq $t1)))
+  (export "e-t1" (type $t2))
+)
+(component
+  (type $t1 (record))
+  (import "t1" (type $t2 (eq $t1)))
+  (import "i" (func $f (result $t2)))
+
+  (export "e-i" (func $f))
+)
+
+;; outer aliases don't work for exports
+(assert_invalid
+  (component
+    (type $t1 (record))
+    (import "t1" (type $t2 (eq $t1)))
+    (component
+      (import "i" (func $f (result $t2)))
+    )
+  )
+  "func not valid to be used as import")
+
+;; outer aliases work for components, modules, and resources
+(component
+  (type $c (component))
+  (type (component
+    (import "c" (component (type $c)))
+  ))
+  (component
+    (import "c" (component (type $c)))
+  )
+  (type $i (instance))
+  (type (component
+    (import "c" (instance (type $i)))
+  ))
+
+  (type $r (resource (rep i32)))
+  (type (component
+    (import "r" (type (eq $r)))
+  ))
+)
+
+;; reexport of an import is fine
+(component
+  (import "r" (func $r))
+  (export "r2" (func $r))
+)
+(component
+  (type $t (record))
+  (import "r" (type $r (eq $t)))
+  (export "r2" (type $r))
+)
+(component
+  (import "r" (instance $r))
+  (export "r2" (instance $r))
+)
+(component
+  (import "r" (type $r (sub resource)))
+  (export "r2" (type $r))
+)
+
+;; TODO - bag of exports can be exported
+(assert_invalid
+  (component
+    (component $A
+      (type $t (record))
+      (export $t2 "t" (type $t))
+      (core module $m (func (export "f")))
+      (core instance $i (instantiate $m))
+      (func $f (result $t2) (canon lift (core func $i "f")))
+
+      (instance (export "i")
+        (export "f" (func $f))
+      )
+    )
+
+    (instance $a (instantiate $A))
+    ;; this component only exports `f`, not the record type that is the result
+    ;; of `f`, so it should be invalid.
+    (export "a" (instance $a "i"))
+  )
+  "instance not valid to be used as export")
+
+;; instance types can be "temporarily invalid", but not if they're attached
+;; to a concrete component
+(component
+  (type (instance
+    (type $t (record))
+    (export "f" (func (param "x" $t)))
+  ))
+)
+(assert_invalid
+  (component
+    (type $i (instance
+      (type $t (record))
+      (type $f (record (field "t" $t)))
+      (export "f" (type (eq $f)))
+    ))
+    (import "f" (instance (type $i)))
```

Comment:
nit: Another way to trigger this same validation condition is, I think, exporting `$i` as a `type` (not the type of an `instance`), which might be a useful test case to add.  I can't recall any other tests for the validity of imported/exported types so, if I haven't missed them, it might be useful to add a few `type` versions of the "instance" validation tests above.